include/tvm/ir/module.h

```diff
@@ -122,6 +122,7 @@ class IRModuleNode : public Object {
     v->Visit("global_var_map_", &global_var_map_);
     v->Visit("global_type_var_map_", &global_type_var_map_);
     v->Visit("source_map", &source_map);
+    v->Visit("attrs", &attrs);
```

Comment:
After having checked :

https://github.com/apache/tvm/blob/9f52e7e3fcb3e607e51b584437857a118fb74609/src/relay/backend/te_compiler.cc#L890-L906

I dont think it is transferred.

Reply:
Hi @manupa-arm, `attrs` were actually added to `IRModule` in https://github.com/apache/tvm/pull/8750 so anything we do now is an iterative improvement on that. Given this series of PRs aims to remove the `Map<Target, IRModule>` entirely (this one removes the `per_target_module` from the interpreter) I don't think we need to ensure the copy happens here but I agree we should have some test coverage when the unified `IRModule` is lowered to ensure it contains all the attributes we've accrued - this should be a follow up when we change the interface from  `Map<Target, IRModule>` to `IRModule` - does that sound reasonable to you?

Reply:
Would there be a PR to remove lowered_funcs from the TE compiler as well ? @electriclilies

Reply:
Its not very clear to me how we can avoid per_target_module throughout the lowering process -- we could push it way down.

Unified IRModule --> per_target_modules (lowered_funcs) --> runtime.Modules

Unless Im missing something here, there will always be a stage that IRModule contains things that gets lowered to single runtime.Module.

Reply:
@manupa-arm I've been pushing the `Map<Target, IRModule>` further and further down the stack, with the goal of eventually removing it. Eventually, I'd like all the APIs to just consume a IRModule where each function is annotated with its target. 

Right now, IRModule's attributes contains all the information needed to construct a LoweredOutput, and I have some conversion functions to extract that info before I stick it into the LoweredOutput. 

You're correct that at this point I'm not copying the attributes into the `per_target_modules` IRModules.  I think that this is OK for now because `per_target_modules` only exists inside a `LoweredOutput`, and the consumers of `LoweredOutput` don't expect there to be attributes on the `per_target_modules`. Additionally, I'm going removing `LoweredOutput` (and `lowered_funcs`) during more follow up work.

Unfortunately,`LoweredOutput` is deeply intertwined with the `build` API, so removing `LoweredOutput`is not a change I can make here. 

My next step will be refactoring the `build` API with the goal of removing `LoweredOutput`. There's also a bunch of other cleanup that needs to be done on the `build` API that will be extensive.

Reply:
Thanks @electriclilies .

In the absense of lowered_funcs, will there be a stage that does 

'IRModule where each function is annotated with its target' --> Array<runtime.Module> then ?
In other words, build API traversing IRModule and emit runtime.Module s in a single shot ?

Reply:
Most likely, although I have not dug into that part of the codebase in depth yet so I can't say for sure. 
The two options that I think are most likely are
1. `build` consuming the IRModule directly, traversing the functions in the IRModule and dealing with each directly (which is what you just mentioned)
2. Right before `build` is invoked, separating the functions in the module by Target (although we wouldn't store them in a `Map<Target, IRModule>`)

So to summarize, we'll either completely remove the data structure that stores functions separated by target, or just push it all the way down to right before `build` is called.

Reply:
So the attrs of IRModule are globally valid to all targets. There I feel conveying them to the codegen might still be beneficial.

For 1,  we would need to expand the build API to pass the attrs

For 2, we can embed them to each IRModule.

The attrs allows a channel that pass through all of the unified lowering process up until the concept of IRModule cease to exist.

If you agree, I feel we should pass them in to the per target IRModule, then later changed in either way we decide to proceed.

Reply:
OK, I can add them to the per target IRModules and the we can figure out what to do with it later.

Reply:
+1 to:
 - preserving module attrs through all passes on the assumption they apply to all targets
 - xfering them when we project a cross-target IRModule to a specific-target IRModule just before the transition into lowering phases

We should probably have a convention that IRModule attrs should be for describing properties of the code and not of the compilation? Since they will appear everywhere it would be tempting to start adding Target-like things there.

Does anyone have thoughts on whether we should try to line up the IRModule and runtime::Module worlds? 
```
                   cross-target         single-target
IR                 IRModule             ?
Executable         ?                    runtime::Module       
```

Reply:
That convention sounds good to me, but one thing we should be cautious of is that the attributes of functions do contain properties of the compilation flow (like targets), and this is inconsistent.

@mbs-octoml By lining up the IRModule and runtime::Module worlds do you mean also moving to a world where runtime::Modules are cross-target? I'm honestly not sure if that's something we want to do. @areusch do you have any thoughts about this?